### ISSUE TITLE
editOnFileSave:true is not working.

### ISSUE THREAD

**Issue by torabu** [stef-levesque/vscode-perforce#110]
VS Code version: 1.18.0
vscode-perforce version: 3.0.0

`editOnFileSave:true` has stopped working for me after updating VS Code to version 1.18. Saving a file will not trigger `p4 edit`, forcing the user to manually call `p4 edit` on the desired file from the terminal.

In my particular case, I have `editOnFileSave` set in User Settings; not sure if that's relevant, but figured it might be worth mentioning.

**Comment by TheHolyWaffle** [comment]
I also experience this issue, which is quite critical.

This is my vscode config:

```
{
    "perforce.deleteOnFileDelete": true,
    "perforce.addOnFileCreate": true,
    "perforce.editOnFileModified": true
}
```

Version of vscode-perforce: 3.0.0
Version of VSCode: 1.18.1

I don't have a .p4config at the root, nor do I set any user, client or port settings. I have also experimented with activationMode set to "always, but that did nothing.

**Comment by TheHolyWaffle** [comment]
Any update on this? It makes it very annoying for us to continue to use VSCode, for now we'll have to fallback to other editors.

**Comment by stef-levesque** [comment]
@jel-massih do you have time to look at this ?

**Comment by galenelias** [comment]
One thing that happens to me quite frequently is if the edit on save doesn't complete fast enough.  The issue I hit is that VSCode itself only gives a small amount of time for the 'pre-save callback' and depending on your server latency, it can take quite a while to edit the file.  @TheHolyWaffle - does this work once on a clean boot of VSCode?  It would be useful to distinguish if it is never working, or just kicking into the 'VSCode disabled the willSave callback'.

Here is the relevant extension documentation (https://code.visualstudio.com/docs/extensionAPI/vscode-ap):
> - there is an overall time budget that all listeners share and if that is exhausted no further listener is called
> - listeners that take a long time or produce errors frequently will not be called anymore
> 
> The current thresholds are 1.5 seconds as overall time budget and a listener can misbehave 3 times before being ignored.

It's also possible it's 'producing an error' and getting the callback disabled that way.

I personally made a bunch of changes to try to make the 'edit' command execute as quickly as possible to avoid this timeout, but it still gets disabled relatively frequently.

**Comment by stef-levesque** [comment]
@TheHolyWaffle if you could try this, on an existing, non-checkout file:
1. `Perforce: Show Output` - `alt+p l`
2. `Perforce: edit - Open an existing file for edit` - `alt+p e`
3. Verify the operation works, that there is no error in the log
4. Check how long it take (as @galenelias noted, 1.5 seconds can be quite fast for some server)

**Comment by galenelias** [comment]
@stef-levesque - I've been running a tweaked copy of this extension where I ended up adding a persistent metadata cache for all the opened files (into Model.ts).  Upon edit I can skip the 'opened' call which is eating up time, and I also can skip the full refresh afterwards (which can be super slow on a many servers and with large changelist - although it doesn't look like it counts against the 1.5s) and instead just do a surgical insert into the cache.  You can also avoid doing a lot of network calls on every file open (one to check if this file is in the depot, and quite often the call to 'print' in provideOriginalResource to get the base version of the file - which you don't need unless the file is opened for edit.

These tweaks do mean you're prone to getting stale behavior if you do source control operations outside of VSCode (until you manually refresh), but for me the performance wins were well worth it (but it was enough of a hack I didn't attempt to get it into shape to integrate it back or anything).  Things like this might help the willSave callback finish in time though (and general performance / depot load).

If this is in fact the problem, other option would be to add an option to basically make the file writable instantly, then do the 'p4 edit' asynchronously.  Again, it leads to the potential to into bad states, but if this is a problem people are hitting, it can be a huge quality of life improvement.

**Comment by stef-levesque** [comment]
Thanks @galenelias for your comments! 😃 

The cache is a good idea, although it really scare me to not be in sync with the server. My personal workflow involves several tools doing p4 operation.

The same goes for preemptively making file writable, I would hate to let the user in a state where it would necessitate a `p4 reconcile`

@joaomoreno  @bpasero @jrieken sry to bothering you, but are you aware of the '1.5 seconds' budget for `onWillSaveTextDocument` ? Is there a way to be more flexible for SCM Providers ?

**Comment by jrieken** [comment]
> @joaomoreno @bpasero @jrieken sry to bothering you, but are you aware of the '1.5 seconds' budget for onWillSaveTextDocument ? 

Sure, that was an important requirement when doing the on-will-save event because we promise our uses that they can always save changes, no matter how extensions are behaving. Not sure how we make this special for SCM providers but ideas have come to make the limit configurable. Needs thinking tho, might need UI because it shouldn't arbitrary long not save without telling the user what's up

**Comment by stef-levesque** [comment]
This is à bit of à special case, as Perforce (and SD) keep files in read-only until the `edit` command make it writable. So, without the extension, the user is not expecting the save to actually complete.

A configurable timeout would make sense, as servers have different latency. As for UI, I would expect a similar sliding/progress-bar as the SCM displays during its operations, although on the TextDocument section.

**Comment by jrieken** [comment]
> I would expect a similar sliding/progress-bar as the SCM displays during its operations, although on the TextDocument section.

We have that but I have doubts that it's enough... Since this will be generic thing I want to give users the change to (1) cancel the operation and proceed with saving and (2) to show the user the extension that's currently doing something...

**Comment by stef-levesque** [comment]
Then a clickable `toolbar` item, with a descriptive hover message might be a better solution.

**Comment by TheHolyWaffle** [comment]
Testing manual edit:
1. Opening a fresh vscode
2. Executing p4 edit command through the bottom toolbar button
3. `p4.exe edit ...` was logged in the output console

Testing the `editOnFileSave: true`:
1. Opening a fresh vscode
2. Modifying a line of code
3. Saving the file through `CTRL+S`
4. No `p4.exe edit ...` was logged in the output console

This tells me that it is not due to the p4 edit command taking up more time than allowed. The command is just simply not executed when trying to save a file, even though `editOnFileSave: true` is set in config.

**Comment by stef-levesque** [comment]
This is then likely an issue with the `FileSystemWatcher` . Can you confirm you are not using multi-root ?

**Comment by TheHolyWaffle** [comment]
I only have a single folder in my project, the same one I worked on before the multi-workspace was released. So no multi-root functionality is used.

**Comment by TheHolyWaffle** [comment]
URL was incorrect, this is what it should be: https://slevesque.gallery.vsassets.io/_apis/public/gallery/publisher/slevesque/extension/perforce/2.2.2/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage

That does the job for now, waiting on the fix...

**Comment by stef-levesque** [comment]
@TheHolyWaffle  This is likely fixed with version 3.1.0
Please update and re-open if it is still happening.
